server/middleware/api.coffee

```diff
@@ -0,0 +1,140 @@
+    throw new errors.Unauthorized('Credentials incorrect.')
+
+  req.client = client
+  next()
+
+  
+postUser = wrap (req, res) ->
+  user = new User({anonymous: false})
+  user.set(_.pick(req.body, 'name', 'email'))
+  user.set('clientCreator', req.client._id)
+  database.validateDoc(user)
+  user = yield user.save()
+  res.status(201).send(user.toObject({req, includedPrivates: INCLUDED_PRIVATES, virtuals: true}))
+  
+  
+getUser = wrap (req, res) ->
+  user = yield database.getDocFromHandle(req, User)
+  if not user
+    throw new errors.NotFound('User not found.')
+
+  unless req.client._id.equals(user.get('clientCreator'))
+    throw new errors.Forbidden('Must have created the user.')
+
+  res.send(user.toObject({req, includedPrivates: INCLUDED_PRIVATES, virtuals: true}))
+  
+  
+postUserOAuthIdentity = wrap (req, res) ->
+  user = yield database.getDocFromHandle(req, User)
+  if not user
+    throw new errors.NotFound('User not found.')
+    
+  unless req.client._id.equals(user.get('clientCreator'))
+    throw new errors.Forbidden('Must have created the user to perform this action.')
+    
+  { provider: providerID, accessToken, code } = req.body or {}
+  unless providerID and (accessToken or code)
+    throw new errors.UnprocessableEntity('Properties "provider" and "accessToken" or "code" required.')
+    
+  if not database.isID(providerID)
+    throw new errors.UnprocessableEntity('"provider" is not a valid id')
+    
+  provider = yield OAuthProvider.findById(providerID)
+  if not provider
+    throw new errors.NotFound('Provider not found.')
+    
+  if code and not accessToken
+    { access_token: accessToken } = yield provider.getTokenWithCode(code)
+    if not accessToken
+      throw new errors.UnprocessableEntity('Code lookup failed')
+
+  userData = yield provider.lookupAccessToken(accessToken)
+  if not userData
+    throw new errors.UnprocessableEntity('User lookup failed')
+    
+  identity = {
+    provider: provider._id
+    id: userData.id
+  }
+
+  otherUser = yield User.findOne({oAuthIdentities: { $elemMatch: identity }})
+  if otherUser
+    throw new errors.Conflict('User already exists with this identity')
+
+  yield user.update({$push: {oAuthIdentities: identity}})
+  oAuthIdentities = user.get('oAuthIdentities') or []
+  oAuthIdentities.push(identity)
+  user.set({oAuthIdentities})
+  res.send(user.toObject({req, includedPrivates: INCLUDED_PRIVATES, virtuals: true}))
+  
+  
+postUserPrepaid = wrap (req, res) ->
+  user = yield database.getDocFromHandle(req, User)
+  if not user
+    throw new errors.NotFound('User not found.')
+
+  unless req.client._id.equals(user.get('clientCreator'))
+    throw new errors.Forbidden('Must have created the user to perform this action.')
+    
+  { endDate } = req.body
+  unless endDate and DATETIME_REGEX.test(endDate)
+    throw new errors.UnprocessableEntity('endDate is not a properly formatted.')
+    
+  # this prepaid goes from when the user 
```

Comment:
Yep, and in finishing it... realized it wasn't handling a use case. [Fixed that, too](https://github.com/codecombat/codecombat/commit/28475ab865f8d3177bb6ccbea3e8fea7f1f99c52).